### PR TITLE
SONiC PAC Testplan HLD

### PR DESCRIPTION
**Pull request by dlakshminarayana**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

### CODE REVIEW

File: docs/testplan/pac/Port_Access_Control.md [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+
+Simple devices like camera or printers which do not support 802.1x authentication can make use of MAB feature where the device gets authenticated based on the device MAC address.
+
+## 1 Test Focus Areas
+
+### 1.1 CLI Testing
+
+	- Verify port authentication can be enabled only on physical interfaces and gets denied on VLAN, Portchannel, PO member ports and sub interfaces.
+	- Verify configured CLI fields are updated properly in respective show commands
+	- Verify PAC configurations/show outputs via REST URI
```
Comment (adyeung) [inline docs/testplan/pac/Port_Access_Control.md:62]: Pls clarify the REST URI test is on SONIC-YANG

File: docs/testplan/pac/Port_Access_Control.md [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+	- Verify non-default max-users per port and check remaining clients are denied.
+	- Verify that when host mode changes, the authenticated clients gets removed and traffic is blocked.
+	- Verify that when authentication order is set to 802.1x, then only 802.1x client allowed to authenticate.
+	- Verify that when authentication order is set to MAB, then 802.1x client authentication is not successful.
+	- Verify MAB client authentication with port-control mode toggle between force authorized/unauthorized and auto.
+	- Verify that 802.1x and MAB client is not authenticated if it's RADIUS assigned VLAN is not available statically on the authenticator switch.
+	- Verify 802.1x and MAB client is not authenticated if RADIUS does not assign a VLAN and the port is configured with tagged VLAN.
+	- Verify a port with Multi-auth mode can have authenticated clients in different radius assigned VLANs.
+	- Verify that 802.1x and MAB client is not authenticated if RADIUS does not assign a VLAN and the port's configured untagged VLAN (Access VLAN) is not available
+
```
Comment (adyeung) [inline docs/testplan/pac/Port_Access_Control.md:90]: I assume these testcases use FreeRADIUS server, pls put a note

**Comment by yxieca** [comment]
@dlakshminarayana are you able to join next test subgroup meeting to present the test plan?

**Comment by adyeung** [comment]
@yxieca the BRCM team will present this after May, will reachout to you for a timeslot when ready

File: docs/testplan/pac/Port_Access_Control.md (hunk 2 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md:95]: Addressed the comments

File: docs/testplan/pac/Port_Access_Control.md (hunk 1 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md]: Addressed the comments

File: docs/testplan/pac/Port_Access_Control.md [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+### Objective
+
+The main objective of this document is to cover the test cases that will be executed for Port authentication methods 802.1x and MAB. Topologies and test cases for testing the feature will be discussed as part of this document.
+
+### Scope
+
+- PAC authentication of hosts on access port
+- This functionality has been tested using the SPyTest framework. In order to emulate 802.1x and MAB clients, traffic generators like Ixia and Spirent will be used. FreeRADIUS is using for User Authentication.
+- Currently, The PAC functionality scoping on platforms AS4630-54PE, Dell N3248TE and Dell N3248PXE.
+- To test the maximum 128 client scalability, the device must have 8GB SSD flash and 4GB DDR4 RAM as minimum requirements.
```
Comment (bingwang-ms) [inline docs/testplan/pac/Port_Access_Control.md:39]: Please remove the 8GB SSD requirement if that's not necessary.

File: docs/testplan/pac/Port_Access_Control.md [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+
```
Comment (bingwang-ms) [inline docs/testplan/pac/Port_Access_Control.md:135]: In Step 16, we don't care the auth type if MAB is disabled.

File: docs/testplan/pac/Port_Access_Control.md [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
...*Steps**      | 1. Enable 802.1x authentication on interface.<br/>2. Configure RADIUS server on the authenticator.<br/>3. Enable authentication host mode to multi-host mode<br/>4. Initiate Authentication from 802.1x supplicant and Verify Authenticator encapsulates EAP packets and sends it to RADIUS server<br/>6. Verify Authenticator moves client to Authenticated state after sending EAP success, once authenticator receives Access-accept from RADIUS server<br/>7. Verify "show authentication clients all" to see the client state<br/>8. Verify client's MAC FDB entry is populated once the client sends traffic after client gets authenticated<br/>9. Verify untagged and tagged traffic gets forwarded from the client's RADIUS assigned VLAN if not port's untagged VLAN after it gets authenticated on the port, check non matching tagged traffic is not allowed.<br/>10. Verify all the subsequent Clients traffic connected to port also gets access and check MAC address gets updated for all those clients<br/>11. Try to authenticate another client and check it is not successful as the port is in multi-host mode.<br/>12. Logoff client1 and verify all the clients connected to the port gets blocked<br/> |
```
Comment (bingwang-ms) [inline docs/testplan/pac/Port_Access_Control.md:166]: Step 11. Authentication is not required.

File: docs/testplan/pac/Port_Access_Control.md [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+| **Test Name**  | **Verify global 802.1x authentication enable/disable.** |
+| **Test Setup** | **Topology1**                            |
+| **Type**       | **Functional**                           |
+| **Steps**      | 1. Enable 802.1x authentication ('config dot1x system-auth-control enable') at global level.<br/>2. Enable 802.1x authentication with single-host on one port and multi-host on another port<br/>3. Initiate Authentication from 802.1x Supplicant and Verify 802.1x aware clients gets authorized on all the ports <br/>4. Verify client gets access to server<br/>5. Disable 802.1x authentication ('config dot1x system-auth-control disable') at global level with 802.1x clients authenticated<br/>6. Verify 802.1x configurations still present under interface and ports move to unauthorized state <br/>7. Verify authenticated 802.1x clients cleared on the authenticator<br/>8. Re-enable global 802.1x authentication and verify all 802.1x clients gets authenticated<br/> |
+
+
+### 3.2.11 Verify enabling re-authentication with different re-authenticate timer and disabling authentication periodic shouldn't allow re-authentication.
+
```
Comment (bingwang-ms) [inline docs/testplan/pac/Port_Access_Control.md:240]: Please also cover the scenario when the auth timer expired.

File: docs/testplan/pac/Port_Access_Control.md [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+| **Test ID**    | **PAC_SCAL_001**                         |
+| -------------- | :--------------------------------------- |
+| **Test Name**  | **Verify 128 max supported 802.1x clients on DUT.**|
+| **Test Setup** | **Topology1**                            |
+| **Type**       | **Functional**                           |
+| **Steps**      | 1. Enable 802.1x authentication globally and at interface level on multiple ports<br/>2. configure-authentication in multi-auth mode<br/>3. Try authentication 128 clients across multiple interfaces(at least 16 clients from one or two interfaces to cover max clients per port) and verify 128 802.1x authenticated clients<br/>4. Verify FDB entries for all 128 clients<br/>5. Verify all client traffic gets allowed after authentication<br/>6. Logoff all the 128 clients and verify clients move to unauthorized state and gets blocked from accessing the server.<br/>7. Reinitiate the clients and check all are authenticated successfully.<br/>8. Log off the clients again and check all are cleared.<br/> |
+
+
+### 3.4.2 Verify 128 max supported MAB clients on DUT.
```
Comment (bingwang-ms) [inline docs/testplan/pac/Port_Access_Control.md:413]: Please include the mix of 802.1x and MAB scenario, meaning the total number of 802.1x and MAB clients is limited to 128.

File: docs/testplan/pac/Port_Access_Control.md [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+
+| **Test ID**    | **PAC_SCAL_002**                         |
+| -------------- | :--------------------------------------- |
+| **Test Name**  | **Verify 128 max supported MAB clients on DUT.**|
+| **Test Setup** | **Topology1**                            |
+| **Type**       | **Functional**                           |
+| **Steps**      | 1. Enable 802.1x authentication globally and at interface level on multiple ports<br/>2. configure-authentication in multi-auth mode<br/>3.Try authentication 128 clients across multiple interfaces and verify 128 MAB authenticated clients <br/>4. Verify FDB entries for all 128 clients<br/>5. Verify all client traffic gets allowed after authentication<br/>6. Logoff all the 128 clients and verify clients move to unauthorized state and gets blocked from accessing the server<br/>7. Reinitiate the clients and check all are authenticated successfully.<br/>8. With max clients authenticated, perform save and reload.<br/>9. Check that after DUT comes up, all clients are authenticated successfully.<br/> |
+
```
Comment (bingwang-ms) [inline docs/testplan/pac/Port_Access_Control.md:421]: Please add a test to cover the MAC move scenario.

File: docs/testplan/pac/Port_Access_Control.md [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+### Objective
+
+The main objective of this document is to cover the test cases that will be executed for Port authentication methods 802.1x and MAB. Topologies and test cases for testing the feature will be discussed as part of this document.
+
+### Scope
+
+- PAC authentication of hosts on access port
+- This functionality has been tested using the SPyTest framework. In order to emulate 802.1x and MAB clients, traffic generators like Ixia and Spirent will be used. FreeRADIUS is using for User Authentication.
+- Currently, The PAC functionality scoping on platforms AS4630-54PE, Dell N3248TE and Dell N3248PXE.
```
Comment (adyeung) [inline docs/testplan/pac/Port_Access_Control.md:38]: Please remove platform references, it's platform independent feature

File: docs/testplan/pac/Port_Access_Control.md (hunk 4 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md:133]: Corrected statement as per review comment.

File: docs/testplan/pac/Port_Access_Control.md (hunk 7 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md:428]: Added new test case

File: docs/testplan/pac/Port_Access_Control.md (hunk 8 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md:446]: Added new test cases.

File: docs/testplan/pac/Port_Access_Control.md (hunk 9 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md]: Removed the statement.

File: docs/testplan/pac/Port_Access_Control.md (hunk 3 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md]: Removed statement.

File: docs/testplan/pac/Port_Access_Control.md (hunk 5 above)
Comment (dlakshminarayana) [reply docs/testplan/pac/Port_Access_Control.md]: Corrected statement as per review comment.

**Review by bingwang-ms — APPROVED** [review]

**Review by adyeung — APPROVED** [review]

**Review by bpar9 — COMMENTED** [review]
The below comments were discussed in testplan review meeting. Adding here for record keeping. Please comment and close if already addressed
- can you verify with macsec enabled on the interface, in multi-host mode?
  if its out of scope, please call out
- can regular ACLs work with PAC in place, can they be verified
- in multi-host/multi-auth modes, can the same client come on different ports? - mac move testcases
- in multi-host mode, can you port-toggle after 1st user logs/authenticates in
- can you verify with config reload

**Comment by dlakshminarayana** [comment]
The below comments were discussed in testplan review meeting. Adding here for record keeping. Please comment and close if already addressed

* can you verify with macsec enabled on the interface, in multi-host mode?
  if its out of scope, please call out
 <Lakshminarayana> PAC and macsec is mutually exclusive. This is out of scope.
* can regular ACLs work with PAC in place, can they be verified
 <Lakshminarayana> PAC and ACLs works independently. Added new test case for this scenario.
* in multi-host/multi-auth modes, can the same client come on different ports? - mac move testcases
 <Lakshminarayana> Added new test case for this scenario.
* in multi-host mode, can you port-toggle after 1st user logs/authenticates in
 <Lakshminarayana> Added new test case for this scenario.
* can you verify with config reload
 <Lakshminarayana> Added new test case for this scenario.

**Comment by adyeung** [comment]
@bpar9 @bingwang-ms please help review the update, if everything is ok pls mark approve

**Review by bingwang-ms — APPROVED** [review]

**Comment by bpar9** [comment]
> @bpar9 @bingwang-ms please help review the update, if everything is ok pls mark approve

thanks lgtm ...approved

**Review by bpar9 — APPROVED** [review]

**Comment by adyeung** [comment]
@yxieca TP has been reviewed and approved, pls help merge

**Review by adyeung — APPROVED** [review]

**Pull request merged**